CI Failure in org.elasticsearch.search.aggregations.metrics.tophits.InternalTopHitsTests#testReduceRandom

`org.elasticsearch.search.aggregations.metrics.tophits.InternalTopHitsTests#testReduceRandom` failed in `6.8`.

It's the first time it failed in 1yr, so not muting yet, but raising the issue for future reference.

log/gradle scan: https://elasticsearch-ci.elastic.co/job/elastic+elasticsearch+6.8+multijob-unix-compatibility/os=debian-9&&immutable/633/console / https://gradle-enterprise.elastic.co/s/22rcz7mz6acn6

Reproduction line:

```
./gradlew ':server:unitTest' \
  -Dtests.seed=7E288303951AD34B \
  -Dtests.class=org.elasticsearch.search.aggregations.metrics.tophits.InternalTopHitsTests \
  -Dtests.method="testReduceRandom" \
  -Dtests.security.manager=true \
  -Dtests.locale=und \
  -Dtests.timezone=America/Nipigon \
  -Dcompiler.java=12 \
  -Druntime.java=8
```

<details><summary>Failure output</summary>

```

Throwable #1: java.lang.AssertionError: Didn't match expected value: |  
-- | --
  | >                           hits: |  
  | >                             hits: |  
  | >                                  0: |  
  | >                                  _id: same [0] |  
  | >                               _score: same [0.69226635] |  
  | >                                _type: same [test] |  
  | >                                  1: |  
  | >                                  _id: same [0] |  
  | >                               _score: same [0.068914354] |  
  | >                                _type: same [test] |  
  | >                                  2: |  
  | >                                  _id: same [0] |  
  | >                               _score: same [0.31542832] |  
  | >                                _type: same [test] |  
  | >                                  3: |  
  | >                                  _id: same [1] |  
  | >                               _score: same [0.37527317] |  
  | >                                _type: same [test] |  
  | >                                  4: |  
  | >                                  _id: same [0] |  
  | >                               _score: same [0.12952006] |  
  | >                                _type: same [test] |  
  | >                                  5: |  
  | >                                  _id: same [0] |  
  | >                               _score: same [0.65314496] |  
  | >                                _type: same [test] |  
  | >                                  6: |  
  | >                                  _id: same [0] |  
  | >                               _score: same [0.19273055] |  
  | >                                _type: same [test] |  
  | >                                  7: |  
  | >                                  _id: same [0] |  
  | >                               _score: same [0.9517262] |  
  | >                                _type: same [test] |  
  | >                                  8: |  
  | >                                  _id: same [0] |  
  | >                               _score: same [0.925982] |  
  | >                                _type: same [test] |  
  | >                                  9: |  
  | >                                  _id: same [0] |  
  | >                               _score: same [0.2846223] |  
  | >                                _type: same [test] |  
  | >                                 10: |  
  | >                                  _id: same [0] |  
  | >                               _score: same [0.93899685] |  
  | >                                _type: same [test] |  
  | >                                 11: |  
  | >                                  _id: same [0] |  
  | >                               _score: same [0.72625506] |  
  | >                                _type: same [test] |  
  | >                                 12: |  
  | >                                  _id: same [0] |  
  | >                               _score: same [0.20685023] |  
  | >                                _type: same [test] |  
  | >                                 13: |  
  | >                                  _id: same [0] |  
  | >                               _score: same [0.31188506] |  
  | >                                _type: same [test] |  
  | >                                 14: |  
  | >                                  _id: same [0] |  
  | >                               _score: same [0.46417063] |  
  | >                                _type: same [test] |  
  | >                                 15: |  
  | >                                  _id: same [0] |  
  | >                               _score: same [0.07432628] |  
  | >                                _type: same [test] |  
  | >                                 16: |  
  | >                                  _id: same [1] |  
  | >                               _score: same [0.87801534] |  
  | >                                _type: same [test] |  
  | >                                 17: |  
  | >                                  _id: same [1] |  
  | >                               _score: same [0.5088928] |  
  | >                                _type: same [test] |  
  | >                                 18: |  
  | >                                  _id: same [0] |  
  | >                               _score: same [0.33007842] |  
  | >                                _type: same [test] |  
  | >                                 19: |  
  | >                                  _id: same [0] |  
  | >                               _score: same [0.81982857] |  
  | >                                _type: same [test] |  
  | >                                 20: |  
  | >                                  _id: same [0] |  
  | >                               _score: same [0.81385064] |  
  | >                                _type: same [test] |  
  | >                                 21: |  
  | >                                  _id: same [0] |  
  | >                               _score: same [0.61173946] |  
  | >                                _type: same [test] |  
  | >                                 22: |  
  | >                                  _id: same [0] |  
  | >                               _score: same [0.1918909] |  
  | >                                _type: same [test] |  
  | >                                 23: |  
  | >                                  _id: same [1] |  
  | >                               _score: same [0.19280052] |  
  | >                                _type: same [test] |  
  | >                                 24: |  
  | >                                  _id: same [1] |  
  | >                               _score: same [0.45122814] |  
  | >                                _type: same [test] |  
  | >                                 25: |  
  | >                                  _id: same [1] |  
  | >                               _score: same [0.4220776] |  
  | >                                _type: same [test] |  
  | >                                 26: |  
  | >                                  _id: same [2] |  
  | >                               _score: same [0.19961417] |  
  | >                                _type: same [test] |  
  | >                                 27: |  
  | >                                  _id: same [1] |  
  | >                               _score: same [0.67341995] |  
  | >                                _type: same [test] |  
  | >                                 28: |  
  | >                                  _id: expected [1] but was [0] |  
  | >                               _score: expected [0.9647093] but was [0.4559121] |  
  | >                                _type: same [test] |  
  | >                        max_score: same [0.9980421] |  
  | >                            total: same [15316587]
```
</details>

--- a/server/src/test/java/org/elasticsearch/search/aggregations/metrics/InternalTopHitsTests.java
+++ b/server/src/test/java/org/elasticsearch/search/aggregations/metrics/InternalTopHitsTests.java
@@ -41,6 +41,7 @@ import org.elasticsearch.search.SearchHit;
 import org.elasticsearch.search.SearchHits;
 import org.elasticsearch.search.aggregations.ParsedAggregation;
 import org.elasticsearch.search.aggregations.pipeline.PipelineAggregator;
+import org.elasticsearch.test.ESTestCase;
 import org.elasticsearch.test.InternalAggregationTestCase;
 import org.elasticsearch.test.NotEqualMessageBuilder;
 
@@ -70,6 +71,12 @@ public class InternalTopHitsTests extends InternalAggregationTestCase<InternalTo
      */
     private SortField[] testInstancesSortFields;
 
+    /**
+     * Collects all generated scores and fields to ensure that all scores are unique. That is necessary for deterministic results
+     */
+    private Set<Float> usedScores = new HashSet<>();
+    private Set<Object> usedFields = new HashSet<>();
+
     @Override
     public void setUp() throws Exception {
         super.setUp();
@@ -88,7 +95,8 @@ public class InternalTopHitsTests extends InternalAggregationTestCase<InternalTo
         SearchHit[] hits = new SearchHit[actualSize];
         Set<Integer> usedDocIds = new HashSet<>();
         for (int i = 0; i < actualSize; i++) {
-            float score = randomFloat();
+            float score = randomValueOtherThanMany(usedScores::contains, ESTestCase::randomFloat);
+            usedScores.add(score);
             maxScore = max(maxScore, score);
             int docId = randomValueOtherThanMany(usedDocIds::contains, () -> between(0, IndexWriter.MAX_DOCS));
             usedDocIds.add(docId);
@@ -97,7 +105,9 @@ public class InternalTopHitsTests extends InternalAggregationTestCase<InternalTo
             if (testInstancesLookSortedByField) {
                 Object[] fields = new Object[testInstancesSortFields.length];
                 for (int f = 0; f < testInstancesSortFields.length; f++) {
-                    fields[f] = randomOfType(testInstancesSortFields[f].getType());
+                    final int ff = f;
+                    fields[f] = randomValueOtherThanMany(usedFields::contains, () -> randomOfType(testInstancesSortFields[ff].getType()));
+                    usedFields.add(fields[f]);
                 }
                 scoreDocs[i] = new FieldDoc(docId, score, fields);
             } else {
@@ -107,10 +117,10 @@ public class InternalTopHitsTests extends InternalAggregationTestCase<InternalTo
             hits[i].score(score);
         }
         int totalHits = between(actualSize, 500000);
+        sort(hits, scoreDocs, scoreDocComparator());
         SearchHits searchHits = new SearchHits(hits, new TotalHits(totalHits, TotalHits.Relation.EQUAL_TO), maxScore);
 
         TopDocs topDocs;
-        Arrays.sort(scoreDocs, scoreDocComparator());
         if (testInstancesLookSortedByField) {
             topDocs = new TopFieldDocs(new TotalHits(totalHits, TotalHits.Relation.EQUAL_TO), scoreDocs, testInstancesSortFields);
         } else {
@@ -120,6 +130,21 @@ public class InternalTopHitsTests extends InternalAggregationTestCase<InternalTo
         TopDocsAndMaxScore topDocsAndMaxScore = new TopDocsAndMaxScore(topDocs, maxScore == Float.NEGATIVE_INFINITY ? Float.NaN : maxScore);
 
         return new InternalTopHits(name, from, requestedSize, topDocsAndMaxScore, searchHits, pipelineAggregators, metaData);
+    }
+
+    /**
+     * Sorts both searchHits and scoreDocs together based on scoreDocComparator()
+     */
+    private void sort(SearchHit[] searchHits, ScoreDoc[] scoreDocs, Comparator<ScoreDoc> comparator) {
+        List<Tuple<SearchHit, ScoreDoc>> hitScores = new ArrayList<>();
+        for (int i = 0; i < searchHits.length; i++) {
+            hitScores.add(new Tuple<>(searchHits[i], scoreDocs[i]));
+        }
+        hitScores.sort((t1, t2) -> comparator.compare(t1.v2(), t2.v2()));
+        for (int i = 0; i < searchHits.length; i++) {
+            searchHits[i] = hitScores.get(i).v1();
+            scoreDocs[i] = hitScores.get(i).v2();
+        }
     }
 
     @Override

